Make all decomps register to Meta, and test it

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #77477
* __->__ #77386

Added an API to let you query if there's already a registration for a function, and then just made `register_decomposition` unconditionally register the meta implementation if there is not already one in core. You can disable it with `disable_meta`.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Make every decomp covering missing Meta register_meta, and test it

Added an API to let you query if there's already a registration for
a function, added a test to check that we register_meta for all
decompositions (except detach, that decomp is wrong lol), and then
update all the necessary sites to make the test pass.

Another strategy is to just unconditionally register_meta if there
isn't a meta registration.  Maybe that's better.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make every decomp covering missing Meta register_meta, and test it"


Added an API to let you query if there's already a registration for
a function, added a test to check that we register_meta for all
decompositions (except detach, that decomp is wrong lol), and then
update all the necessary sites to make the test pass.

Another strategy is to just unconditionally register_meta if there
isn't a meta registration.  Maybe that's better.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make all decomps register to Meta, and test it"


Added an API to let you query if there's already a registration for a function, and then just made `register_decomposition` unconditionally register the meta implementation if there is not already one in core. You can disable it with `disable_meta`.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -14460,7 +14461,7 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
            supports_fwgrad_bwgrad=True,
            gradcheck_wrapper=lambda *args, **kwargs: gradcheck_wrapper_triangular_input(*args, idx=1, **kwargs),
-           decorators=[skipCUDAIfNoMagma],
+           decorators=[skipCUDAIfNoMagma, skipCPUIfNoLapack],
```
Nice update

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -14705,6 +14706,7 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
            dtypes=floating_and_complex_types(),
            sample_inputs_func=sample_inputs_linalg_solve_triangular,
            supports_fwgrad_bwgrad=True,
+           skips=(skipCPUIfNoLapack,),
```
Nice skip

mruberry reviewed (approved):
LGTM!

Chillee commented on test/test_meta.py:
```diff
@@ -331,7 +332,6 @@
     torch.nn.functional.adaptive_avg_pool2d,  # MISSING aten::_adaptive_avg_pool2d
     torch.nn.functional.adaptive_avg_pool3d,  # MISSING aten::_adaptive_avg_pool3d
     torch.nn.functional.batch_norm,  # MISSING aten::native_batch_norm
-    torch.nn.functional.binary_cross_entropy,  # MISSING aten::binary_cross_entropy
```
Are the `# missing` parts autogenerated btw?

ezyang replied:
yup

Chillee commented:
I'm a little ... surprised the tests all pass? Most of the decomps haven't had that much attention paid to strides so far. Perhaps the set of ops for which we've written decompositions mostly have trivial stride propagation.

ezyang commented:
> I'm a little ... surprised the tests all pass? Most of the decomps haven't had that much attention paid to strides so far. Perhaps the set of ops for which we've written decompositions mostly have trivial stride propagation.

There's a testing problem I'm working on, which is that no backwards is being tested for meta.

ezyang committed:
Update on "Make all decomps register to Meta, and test it"


Added an API to let you query if there's already a registration for a function, and then just made `register_decomposition` unconditionally register the meta implementation if there is not already one in core. You can disable it with `disable_meta`.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make all decomps register to Meta, and test it"


Added an API to let you query if there's already a registration for a function, and then just made `register_decomposition` unconditionally register the meta implementation if there is not already one in core. You can disable it with `disable_meta`.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
folded into https://github.com/pytorch/pytorch/pull/77477
